src/Compiler/Microsoft.AspNetCore.Razor.Language/src/DefaultRequiredAttributeDescriptorBuilder.cs

```diff
@@ -1,87 +1,79 @@
+using System.Collections.Immutable;
+using Microsoft.AspNetCore.Razor.PooledObjects;
 
 namespace Microsoft.AspNetCore.Razor.Language;
 
-internal class DefaultRequiredAttributeDescriptorBuilder : RequiredAttributeDescriptorBuilder
+internal class DefaultRequiredAttributeDescriptorBuilder : RequiredAttributeDescriptorBuilder, IBuilder<RequiredAttributeDescriptor>
 {
     private readonly DefaultTagMatchingRuleDescriptorBuilder _parent;
-    private RazorDiagnosticCollection _diagnostics;
-    private readonly Dictionary<string, string> _metadata = new Dictionary<string, string>();
+    private RazorDiagnosticCollection? _diagnostics;
+    private readonly ImmutableDictionary<string, string>.Builder _metadata;
 
     public DefaultRequiredAttributeDescriptorBuilder(DefaultTagMatchingRuleDescriptorBuilder parent)
     {
         _parent = parent;
+        _metadata = ImmutableDictionary.CreateBuilder<string, string>();
     }
 
-    public override string Name { get; set; }
+    public override string? Name { get; set; }
 
     public override RequiredAttributeDescriptor.NameComparisonMode NameComparisonMode { get; set; }
 
-    public override string Value { get; set; }
+    public override string? Value { get; set; }
 
     public override RequiredAttributeDescriptor.ValueComparisonMode ValueComparisonMode { get; set; }
 
-    public override RazorDiagnosticCollection Diagnostics
-    {
-        get
-        {
-            if (_diagnostics == null)
-            {
-                _diagnostics = new RazorDiagnosticCollection();
-            }
-
-            return _diagnostics;
-        }
-    }
+    public override RazorDiagnosticCollection Diagnostics => _diagnostics ??= new RazorDiagnosticCollection();
 
     public override IDictionary<string, string> Metadata => _metadata;
 
     internal bool CaseSensitive => _parent.CaseSensitive;
 
     public RequiredAttributeDescriptor Build()
     {
-        var diagnostics = Validate();
-        if (_diagnostics != null)
+        var diagnostics = new PooledHashSet<RazorDiagnostic>();
+        try
         {
-            diagnostics ??= new();
+            Validate(ref diagnostics);
+
             diagnostics.UnionWith(_diagnostics);
-        }
 
-        var displayName = GetDisplayName();
-        var rule = new DefaultRequiredAttributeDescriptor(
-            Name,
-            NameComparisonMode,
-            CaseSensitive,
-            Value,
-            ValueComparisonMode,
-            displayName,
-            diagnostics?.ToArray() ?? Array.Empty<RazorDiagnostic>(),
-            new Dictionary<string, string>(Metadata));
-
-        return rule;
+            var displayName = GetDisplayName();
+
+            var descriptor = new DefaultRequiredAttributeDescriptor(
+                Name,
+                NameComparisonMode,
+                CaseSensitive,
+                Value,
+                ValueComparisonMode,
+                displayName,
+                diagnostics.ToArray(),
```

Comment:
This seems to lead to non-deterministic diagnostics. The before code has the same property so this isn't changing anything but this still seems surprising. Is the non-determinism normalized out later somewhere?

Reply:
I agree that it's weird. I thought about it in the moment and decided that not to deviate from the existing behavior. If we want to make the Razor compiler and tag helper descriptors non-deterministic, I think we've got a lot more work cut out for us.

Reply:
FWIW, the non-determinism is not normalized elsewhere.